ledger/src/blockstore/blockstore_purge.rs

```diff
@@ -391,18 +396,32 @@ impl Blockstore {
 
         for slot in from_slot..=to_slot {
             let primary_indexes = slot_indexes(slot);
-            if primary_indexes.is_empty() {
-                continue;
-            }
+            let delete_new_column_key =
+                primary_indexes.is_empty() || (slot == cmp::max(index0.max_slot, index1.max_slot));
 
             let slot_entries = self.get_any_valid_slot_entries(slot, 0);
             let transactions = slot_entries
                 .into_iter()
                 .flat_map(|entry| entry.transactions);
-            for transaction in transactions {
+            for (i, transaction) in transactions.enumerate() {
                 if let Some(&signature) = transaction.signatures.get(0) {
+                    if delete_new_column_key {
+                        batch.delete::<cf::TransactionStatus>((signature, slot))?;
+                        batch.delete::<cf::TransactionMemos>((signature, slot))?;
+                    }
```

Comment:
Initially, I thought it should be the case that we only have new OR old key format for a given slot/transaction, but now I think we could have a mix. Consider the following sequence:
- Node is running with version that is writing new key formats
- Node crashes midway through replay of slot `X`
- Operator reverts software to version that writes old key formats
- Node catches up / replays and eventually replays slot `X`

I think slot `X` would now have a mix of old/new key formats. So, maybe we do need to explicitly check for both old and new key formats for each transaction. I started this comment hoping we could change to something like this, but per above, don't think we can go this way.
```rust
if delete_new_column_key {
    // delete new key format
} else {
    // delete old key format
}

Reply:
Yes, I had originally written an if/else, but came to the same conclusion. Even the first slot written in the new key format could co-exist.

Yes, I think to be most efficient with entries in the delete batch, we do need to check each key format. I have had some more thoughts about how to do this. But I'd like to do that in a separate PR.
For the time being, I will change this to always delete the new-format key, and optionally also try to delete the deprecated formats.

Reply:
Faulty logic removed in [Remove faulty delete_new_column_key logic](https://github.com/solana-labs/solana/pull/33419/commits/4b68c044f89b43460550e82031cbd49da6fd09e7)